File: stream-dynamo-consumer/src/database/dynamo.writer.ts
import dynamoose from "dynamoose";
import { Model } from "dynamoose/dist/Model";
import { SchemaDefinition } from "dynamoose/dist/Schema";
import { DynamoDB, DynamoDBClientConfig } from "@aws-sdk/client-dynamodb";

export default class DynamoWriter<T> {
  private ddb: DynamoDB;
  private model: Model<any>;

  constructor(readonly schemaDefinition: SchemaDefinition) {
    const host = process.env.DYNAMO_HOST || "localhost";
    const port = process.env.DYNAMO_PORT || "8000";
    const table = process.env.DYNAMO_TABLE || "ChatMessages";

    // DynamoDB 클라이언트 설정
    const config: DynamoDBClientConfig = {
      credentials: {
        accessKeyId: "fakeKey", // 로컬에서 사용할 가짜 자격 증명
        secretAccessKey: "fakeSecretKey",
      },
      endpoint: `http://${host}:${port}`,
      region: "local", // 로컬 설정
    };

    // DynamoDB 클라이언트 초기화
    this.ddb = new dynamoose.aws.ddb.DynamoDB(config);

    // Dynamoose에 DynamoDB 클라이언트 설정
    dynamoose.aws.ddb.set(this.ddb);

    // 스키마 정의
    const schema = new dynamoose.Schema(schemaDefinition);

    // 모델 초기화
    this.model = dynamoose.model(table, schema);
  }

  // 데이터 업데이트
  async updateChatMessage(eventId: string, content: T & { createdAt?: Date }) {
    try {
      await this.model.update(
        { eventId }, // Partition Key
        { ...content } // 업데이트할 필드
      );
    } catch (error) {
      console.error("Error updating message:", error);
    }
  }
}

File: sink/dynamo-consumer/src/main/scala/database/DynamoWriter.scala
package consumers.dynamo.sink
package database

import java.net.URI
import java.time.Instant
import scala.jdk.CollectionConverters._

import software.amazon.awssdk.auth.credentials.{AwsBasicCredentials, StaticCredentialsProvider}
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.dynamodb.model._
import software.amazon.awssdk.services.dynamodb.DynamoDbClient

import consumers.shared.config.Env.{DYNAMO_HOST, DYNAMO_PORT, DYNAMO_TABLE}
import consumers.dynamo.sink.database.given

class DynamoWriter[T <: Product: DynamoMappable]() {
  private val endpoint = s"http://$DYNAMO_HOST:$DYNAMO_PORT"

  // DynamoDB 클라이언트 설정
  private val client: DynamoDbClient = DynamoDbClient
    .builder()
    .endpointOverride(URI.create(endpoint))
    .region(Region.US_WEST_2)
    .credentialsProvider(
      StaticCredentialsProvider.create(
        AwsBasicCredentials.create(
          "fakeKey", // 로컬에서 사용할 가짜 자격 증명
          "fakeSecretKey"
        )
      )
    )
    .build()

  // 데이터 업데이트 함수
  def updateChatMessage(eventId: String, chat: T): Unit = {
    try {
      // DynamoMappable을 이용해 ChatMessage를 Map으로 변환
      val content = summon[DynamoMappable[T]].toAttributes(chat)

      // DynamoDB 업데이트 작업에서 사용할 SQL-like 업데이트 표현식
      val updateExpr = content.keys.zipWithIndex
        .map { case (k, i) => s"#k$i = :v$i" }
        .mkString(", ")
      val exprAttrNames = content.keys.zipWithIndex
        .map { case (k, i) => s"#k$i" -> k }
        .toMap
        .asJava
      val exprAttrValues = content.values.zipWithIndex
        .map { case (v, i) => s":v$i" -> v }
        .toMap
        .asJava

      // DynamoDB의 UpdateItemRequest 객체 생성
      val updateRequest = UpdateItemRequest
        .builder()
        .tableName(DYNAMO_TABLE)
        .key(Map("eventId" -> AttributeValue.builder().s(eventId).build()).asJava)
        .updateExpression(s"SET $updateExpr") // 업데이트 표현식
        .expressionAttributeNames(exprAttrNames) // 속성 이름 매핑
        .expressionAttributeValues(exprAttrValues) // 속성 값 매핑
        .build()

      // DynamoDB에 업데이트 요청
      client.updateItem(updateRequest)
    } catch {
      case e: Exception =>
        println(s"Error updating message: ${e.getMessage}")
    }
  }
}
